openhcl/page_pool_alloc/src/lib.rs

```diff
@@ -23,6 +23,261 @@ use std::sync::Arc;
+            #[mesh(1)]
+            base_pfn: u64,
+            #[mesh(2)]
+            pfn_bias: u64,
+            #[mesh(3)]
+            size_pages: u64,
+            #[mesh(4)]
+            device_id: usize,
+            #[mesh(5)]
+            tag: String,
+        },
+    }
+
+    impl From<State> for InnerState {
+        fn from(state: State) -> Self {
+            match state {
+                State::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                } => InnerState::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                },
+                State::Allocated {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => InnerState::Allocated {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+                State::AllocatedPendingRestore { .. } => {
+                    panic!("should not save AllocatedPendingRestore")
+                }
+                State::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => InnerState::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+            }
+        }
+    }
+
+    impl From<InnerState> for State {
+        fn from(state: InnerState) -> Self {
+            match state {
+                InnerState::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                } => State::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                },
+                InnerState::Allocated {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => State::AllocatedPendingRestore {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+                InnerState::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => State::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+            }
+        }
+    }
+
+    // TODO: It seems unfortunate to define this type which is the same as
```

Comment:
So you're saying that we cannot import from vm_topology here?

Reply:
we cannot use it as part of saved state, because it's not defined as a protobuf derive struct, unlike `MemoryRange`

Reply:
Do we really need to save the node anyway? That doesn't seem important. If it changes (maybe due to some bug or bug fix), so what?

Reply:
Eh i think that's fair we can just not save the vnode.